urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlib/utilities/InvolutionUtilities.java

```diff
@@ -0,0 +1,96 @@
+import java.util.List;
+import permlib.PermUtilities;
+import permlib.Permutation;
+import permlib.property.PermProperty;
+
+/**
+ * Utility functions for dealing with involutions.
+ * 
+ * @author Michael Albert
+ */
+public class InvolutionUtilities {
+
+    /**
+     * Computes the direct, i.e. one or two point extensions of an involution 
+     * that are also involutions. Returns an empty list if the input
+     * permutation is not an involution.
+     * 
+     * @param p the involution
+     * @return the extensions
+     */
+    public static List<Permutation> directInvolutionExtensions(Permutation p) {
+        LinkedList<Permutation> result = new LinkedList<Permutation>();
+        if (!isInvolution(p)) {
+            return result;
+        }
+        result.add(PermUtilities.insert(p, p.length(), p.length()));
+        for (int k = 0; k <= p.length(); k++) {
+            Permutation q = PermUtilities.insert(p, p.length(), k);
+            result.add(PermUtilities.insert(q, k, p.length() + 1));
+        }
+        return result;
+    }
+
+    /**
+     * Computes the direct, i.e. one or two point extensions of an involution 
+     * that are also involutions and satisfy a given property. Returns an empty
+     * list if the input permutation is not an involution.
+     * 
+     * @param p an involution
+     * @param prop a property
+     * @return the extensions of the involution that satisfy the property
+     */
+    public static List<Permutation> directInvolutionExtensions(Permutation p, PermProperty prop) {
+        LinkedList<Permutation> result = new LinkedList<Permutation>();
+        if (!isInvolution(p)) {
+            return result;
+        }
+        Permutation q = PermUtilities.insert(p, p.length(), p.length());
+        if (prop.isSatisfiedBy(q)) {
+            result.add(q);
+        }
+        for (int k = 0; k <= p.length(); k++) {
+            q = PermUtilities.insert(p, p.length(), k);
+            q = PermUtilities.insert(q, k, p.length() + 1);
+            if (prop.isSatisfiedBy(q)) {
+                result.add(q);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Tests whether a given permutation is an involution or not.
+     * 
+     * @param p the permutation to test
+     * @return true if the permutation is an involution, false otherwise
+     */
+    public static boolean isInvolution(Permutation p) {
+        for (int i = 0; i < p.length(); i++) {
+            if (p.elements[p.elements[i]] != i) {
+                return false;
+            }
+        }
+        return true;
+    }
+    
+    /**
+     * Inserts a point or pair in a (presumed) involution to form a new involution.
+     * @param p the involution
+     * @param i the position of the point
+     * @param j the value of the point
+     * @return the result of the insertion
+     */
+    public static Permutation insert(Permutation p, int i, int j) {
+        Permutation result = p.insert(i,j);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;p&#39; (index: 0 | type: Permutation). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;i&#39;, &#39;j&#39;. Other methods validating the parameter: &#39;directInvolutionExtensions&#39;, &#39;directInvolutionExtensions&#39;. The same parameter type is validated here for example: urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlab/examples/EdelmanGreene.java:111. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).